stock_valuation_fifo_lot/models/product.py

```diff
@@ -0,0 +1,112 @@
+# Copyright 2022 Ecosoft Co., Ltd (https://ecosoft.co.th)
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl.html)
+
+from odoo import models
+from odoo.tools import float_is_zero
+
+
+class ProductProduct(models.Model):
+    _inherit = "product.product"
+
+    def _get_all_candidates(self, company):
+        all_candidates = (
+            self.env["stock.valuation.layer"]
+            .sudo()
+            .search(
+                [
+                    ("product_id", "=", self.id),
+                    ("remaining_qty", ">", 0),
+                    ("company_id", "=", company.id),
+                ]
+            )
+            .sorted(lambda l: l.lot_ids.create_date)
+        )
+        return all_candidates
+
+    def _run_fifo(self, quantity, company):
+        self.ensure_one()
+        move_id = self._context.get("used_in_move_id")
+        if self.tracking == "none" or not move_id:
+            vals = super()._run_fifo(quantity, company)
+        else:
+            move = self.env["stock.move"].browse(move_id)
+            move_lines = move._get_out_move_lines()
+            tmp_value = 0
+            tmp_remaining_qty = 0
+            for move_line in move_lines:
+                # Find back incoming stock valuation layers
+                # (called candidates here) to value `quantity`.
+                qty_to_take_on_candidates = move_line.product_uom_id._compute_quantity(
+                    move_line.qty_done, move.product_id.uom_id
+                )
+                candidates = (
+                    self.env["stock.valuation.layer"]
+                    .sudo()
+                    .search(
+                        [
+                            ("product_id", "=", self.id),
+                            ("remaining_qty", ">", 0),
+                            ("company_id", "=", company.id),
+                        ]
+                    )
```

Comment:
... so that the method can be reused here.
```suggestion
                    self._get_all_candidates(company)
```